[RDY] clint: detect MAYBE and recommend TriState

janlazo opened the pull request:
Follow-up from patching 8.0.0831 in https://github.com/neovim/neovim/pull/8742.

janlazo committed:
clint: detect MAYBE and recommend TriState

Vim uses MAYBE for 3-value boolean with FALSE/TRUE/MAYBE.
Use TriState type instead to restrict to 3 values, kFalse/kTrue/kNone.

janlazo commented:
This will take awhile.

```
src/nvim/diff.c:53:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/diff.c:742:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/diff.c:758:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/edit.c:598:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/edit.c:7177:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/ex_cmds.c:5056:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/ex_cmds.c:5151:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/ex_cmds.c:5221:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/ex_cmds.c:5235:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/ex_cmds.c:5237:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/ex_docmd.c:7427:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:53:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:652:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:790:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:1343:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:1344:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:1530:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:1534:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:1556:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:1584:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:2364:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:2457:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:2569:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:2570:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/hardcopy.c:675:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/hardcopy.c:676:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/hardcopy.c:677:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/hardcopy.c:2858:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/menu.c:163:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/menu.c:166:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/menu.c:171:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/normal.c:2018:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/normal.c:2116:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/ops.c:5378:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/screen.c:832:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/search.c:1757:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/search.c:1931:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/search.c:1942:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/search.c:1955:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/state.c:82:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/globals.h:957:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/vim.h:181:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
Total errors found: 42
```

janlazo committed:
diff: refactor diff_a_works to use TriState

janlazo committed:
fold: fold_T.fd_small is TriState

janlazo committed:
menu: enable in ex_menu() is TriState

janlazo committed:
ex_cmds: refactor utf8 variables to TriState

- utf8 in helptags_one()
- this_utf in fix_help_buffer()

janlazo committed:
edit: dont_sync_undo is TriState

janlazo committed:
search: start_in_quotes in findmatchlimit is TriState

janlazo committed:
search: refactor variables in findmatchlimit()

Combine declaration and initialization.
Refactor int variables, that use TRUE/FALSE macros (not MAYBE), to bool.

janlazo committed:
fold: check_closed() returns bool

Update affected variables (ie. had_folded).
Add const to params to restrict usage.

TODO: refactor win_T.w_lines[idx].wl_folded from char to bool

janlazo committed:
fold: lineFolded() is bool

janlazo committed:
screen: screen_cleared is TriState

janlazo commented:
I'll settle with this to avoid including patches here.

```
src/nvim/ex_docmd.c:7427:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/fold.c:1520:  Use true instead of TRUE.  [readability/bool] [4]
src/nvim/fold.c:1530:  Use true instead of TRUE.  [readability/bool] [4]
src/nvim/fold.c:2370:  Use true instead of TRUE.  [readability/bool] [4]
src/nvim/fold.c:2372:  Use true instead of TRUE.  [readability/bool] [4]
src/nvim/fold.c:2459:  Use true instead of TRUE.  [readability/bool] [4]
src/nvim/hardcopy.c:675:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/hardcopy.c:676:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/hardcopy.c:677:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/hardcopy.c:2858:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/normal.c:2018:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/normal.c:2116:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/ops.c:5378:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/state.c:82:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/globals.h:957:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
src/nvim/vim.h:181:  Use kNONE from TriState instead of MAYBE.  [readability/bool] [4]
Total errors found: 16
```

janlazo committed:
fold: finish in foldUpdateIEMSRecurse() is bool

janlazo committed:
fold: fold_changed is bool

janlazo committed:
fold: recursive in deleteFoldEntry() is bool

jamessan commented:
Note that the values of `TriState` were specifically chosen to still allow it to be used in a boolean context, if the condition simply cares about relation to `kFalse`.  That is, there's no need to change `if (foo)` to `if (foo != kFalse)` or `if (!foo)` to `if (foo == kFalse)`.

janlazo commented:
I made them explicit, diverging from Vim, because we're using `kNone`, not `MAYBE`, and "none" implies a falsy value even if it is a non-zero value and evaluates to true.

justinmk commented:
I'm not sure we should diverge from vim here, in the near-term, if it will complicate vim-patch porting.

janlazo commented on src/nvim/diff.c:
```diff
@@ -830,7 +830,7 @@ static void diff_file(const char *const tmp_orig, const char *const tmp_new,
      * non-zero when differences have been found. */
     vim_snprintf(cmd, len, "diff %s%s%s%s%s %s",
-                 diff_a_works ? "-a " : "",
+                 diff_a_works == kFalse ? "" : "-a ",
```
This is a "revert" (upstream has same line (swap `kFalse` with `FALSE`).

janlazo commented on src/nvim/diff.c:
```diff
@@ -739,23 +739,23 @@ void ex_diffupdate(exarg_T *eap)
 
     // If "-a" works break here, otherwise retry without "-a".
-    if (ok) {
+    if (ok != kFalse) {
```
Same as latest code from upstream (checked yesterday).

janlazo commented on src/nvim/diff.c:
```diff
@@ -739,23 +739,23 @@ void ex_diffupdate(exarg_T *eap)
   }
 
-  if (!ok) {
+  if (ok == kFalse) {
```
Hasn't changed in upstream (checked yesterday).

janlazo commented on src/nvim/fold.c:
```diff
@@ -274,14 +274,11 @@ int foldLevel(linenr_T lnum)
+// Return false if line is not folded.
+bool lineFolded(win_T *const win, const linenr_T lnum)
 {
   return foldedCount(win, lnum, NULL) != 0;
```
This expression looks boolean to me and doesn't return `MAYBE` so i changed this bool and updated affected code.

janlazo committed:
globals: virtual_op is TriState

janlazo committed:
fold: use_level,maybe_small are bool

use_levelp, maybe_smallp in check_closed() are bool ptrs.

janlazo commented:
My other PRs aren't getting merged so I'll work on `hardcopy.c` in the meantime so that the lint build passes.

janlazo committed:
hardcopy: bold,italic,underline are TriState

janlazo committed:
hardcopy: refactor mch_print_start_line()

Refactor the following affected functions:
- prt_line_number()
- prt_header()

These use prt_set_font(), refactored in 872a76f3dac14fdcb7dc7b34cecd3acb4370d56f for TriState.

janlazo commented:
Any issues with merging this?

justinmk commented:
I was hesitant because it seems risky, but after another look I feel better about it. Will review in a bit.

janlazo commented:
@justinmk Still risky?

janlazo commented:
Ping

janlazo committed:
ex_cmds: add const to fix_help_buffer() variables

janlazo committed:
ex_cmds: add const to helptags_one() variables

janlazo committed:
fold: add const to hasFoldingWin() variables

cache is bool so update callers to pass true/false, not TRUE/FALSE.

janlazo committed:
fold: add const to deleteFold() variables

Declare and initialize variables on same line if possible.

janlazo committed:
fold: add const to foldMoveTo() variables

Declare and initialize variables on same line if possible.

janlazo committed:
fold: declare and init vars in deleteFoldEntry()

Use sizeof() on pointer value, not explicit type, for memmove.

janlazo committed:
fold: add const to checkSmall() variables

janlazo committed:
fold: add const to foldUpdateIEMS() variables

janlazo committed:
fold: add const to foldUpdateIEMSRecurse() vars

janlazo committed:
fold: add const to foldSplit() variables

janlazo committed:
misc: refactor plines_win{,_nofill}()

Add const to params and variables (declare and init on same line).
winheight (param) is bool so replace TRUE/FALSE macros with true/false.

janlazo commented:
I overdid the style changes here and in  #8799 (to pass the time :neutral_face: ) but I want this, #8799, and #8744 merged before continuing on my other vim-patch PRs so that I don't have to rebase on top of them to avoid conflicts later on. The lint build fails because of the new warning introduced and the only error it reports is the `#define` for `MAYBE`.

justinmk commented on src/nvim/fold.c:
```diff
@@ -1505,36 +1486,40 @@ static int getDeepestNestingRecurse(garray_T *gap)
+    fold_T *const fp,
+    bool *const use_levelp,     // true: outer fold had FD_LEVEL
+    const int level,            // folding depth
+    bool *const maybe_smallp,   // true: outer this had fd_small == kNone
```
It looks like you `const`'d the pointer value rather than the pointer contents. Why?

janlazo replied:
Adding `const` to pointer value ensures that it doesn't change in the function and it's safe to pass to functions with non-const pointer params without typecast. C is mutable by default, including the params, so adding `const` gives some assurance that it doesn't change midway in a long function.

bfredl replied:
Adding const to every local binding that happens to not change will add a lot of linenoise to the implementation. Overly long functions (so long that local vars are hard to keep track of) are the exception, they at least shouldn't determine how we write reasonably sized functions. (Didn't check this function because of on mobile)
